x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2777 @@
+
+This type of migration will be in charge of updating saved object fields in Elasticsearch from the legacy form to the new form, and will be performed on all write operations of the rules (see list below). This means that we need to perform this type of migration of rules whenever any endpoint operation that writes/updates a rule saved object is called.
+
+This migration strategy means that rules will be migrated incrementally, and that both non-migrated and migrated rules will coexist for an indeterminate amount of time. Therefore we have to maintain backwards compatibility with non-migrated rules.
+
+The migration logic should take place within the handler logic of all endpoints that carry out write/update endpoint operations, and the endpoint should return the already-migrated rule(s).
+
+Endpoints that perform migration-on-write either fetch the rule before updating it, or send the rule's params as part of the body in the request. Therefore, we can apply migraton logic to them -as described below- before saving the rules to Elasticsearch.
+
+### Technical implementation of migration-on-write
+
+The logic for the migration of the rule saved objects, which means the determination of the `immutable` and `rule_source` fields before writing to ES, might differ depending on the action being performed by the user.
+
+Let's see all possible use cases that will require migration-on-write, the endpoints that they apply to, and the expected resulting migrated field, based on the action and their input.
+
+#### Updating and upgrading rules
+
+Updating rules can currently be performed via five endpoints:
+
+- **Update Rule** - `PUT /rules`
+- **Patch Rule** - `PATCH /rules`
+- **Bulk Update Rules** - `PUT /rules/_bulk_update`
+- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`
+- **Bulk Actions** - `POST /rules/_bulk_action`: with **bulk edit** action
+
+Upgrading prebuilt rules to their newer version is done by two endpoints:
+
+- **(LEGACY) Install Prebuilt Rules And Timelines** - `PUT /rules/prepackaged`
+- **Perform Rule Upgrade** - `POST /prebuilt_rules/upgrade/_perform` (Internal)
+
+The legacy endpoint does not allow for customization of fields during the upgrade, but the new rule upgrade customization endpoint does. 
+
+Additionally:
+
+- **Bulk Actions** - `POST /rules/_bulk_action`: with **duplicate** action
+
+will perform migration but does not allow for customization during the duplication process.
+
+So we can analyze the expected outputs of the migration of all these 8 endpoints together.
+
+The resulting values for `immutable` and `rule_source` when calling these endpoints, and the migration being performed in the background, should be as follows:
+
+<table>
+  <thead>
+    <tr>
+      <th>Migration use case</th>
+      <th>Current value of <pre>rule_source</pre></th>
+      <th>Current value of <pre>immutable</pre></th>
+      <th>Any field diverges <br> from base version after update?</th>
+      <th>Results</th>
```

Comment:
Nit: In the "results" column we show a json containing the rule source and immutable parameters, but for the currently stored values we have two columns. I think we could improve readability of the table if we used a single column for the currently stored value.

```suggestion
      <th>Currently stored</th>
      <th>Any field diverges <br> from base version after update?</th>
      <th>Result to be stored</th>
```

Reply:
Fixed.

One note for when we discuss this: this examples consider that `immutable` is not deleted when the migration-on-write is done. As you mentioned in another comment, I think now it should actually be deleted. Let's discuss tomorrow this, but I can remove the `immutable` field where the rule was already migrated in this table.